server/server.go

```diff
@@ -42,72 +55,362 @@ func HandleWebsocketConnection(serverHTTPResponse http.ResponseWriter, clientHTT
-	if protobufAuthBytesErr != nil {
+	messageToClientBytes, messageToClientBytesErr := proto.Marshal(&messageToClient)
+	if messageToClientBytesErr != nil {
 		// If there is an error while trying to perform protobuf marshaling, terminate the connection
-		return websocket.BinaryMessage, protobufAuthBytesErr
-	}
+		return "", messageToClientBytesErr
+	} // if messageToClientBytesErr != nil
 	// Send the protobuf data to the client containing the sequence of bytes he needs to sign
-	websocketConnection.WriteMessage(websocket.BinaryMessage, protobufAuthBytes)
+	websocketConnection.WriteMessage(websocket.BinaryMessage, messageToClientBytes)
 	// Read the response from the client which should contain his IdenetityPublicKey and the signed byte sequence
-	_, clientMessage, readMessageErr := websocketConnection.ReadMessage()
+	_, messageFromClientProto, readMessageErr := websocketConnection.ReadMessage()
 	if readMessageErr != nil {
 		// If there is an error while reading the response from the client, terminate the connection
-		return websocket.BinaryMessage, readMessageErr
+		return "", readMessageErr
 	}
 	// Unmarshal the response from the client into our protobuf Auth structure
-	protoUnmarshalErr := proto.Unmarshal(clientMessage, &protobufAuth)
+	protoUnmarshalErr := proto.Unmarshal(messageFromClientProto, &messageFromClient)
 	if protoUnmarshalErr != nil {
 		// If there is an error while unmarshaling the response from the client, terminate the connection
-		return websocket.BinaryMessage, protoUnmarshalErr
+		return "", protoUnmarshalErr
 	}
 	// Create a [32]byte{} identityPublicKey to satisfy ed25519.Verify() type requirements
 	identityPublicKey := [32]byte{}
 	// Create a [64]byte{} signedByteSequenceFromClient to satisfy ed25519.Verify() type requirements
 	signedByteSequenceFromClient := [64]byte{}
+	// Create a string representation of the Identity Public Key
+	identityPublicKeyString := string(messageFromClient.Response.Auth.IdentityPublicKey)
 	// Decode the hex representation of the identityPublicKey here as transmitting it over the network without hex encoding can cause issues with some clients
-	identityPublicKeyDecoded, identityPublicKeyDecodedErr := hex.DecodeString(string(protobufAuth.GetIdentityPublicKey()))
+	identityPublicKeyDecoded, identityPublicKeyDecodedErr := hex.DecodeString(identityPublicKeyString)
 	if identityPublicKeyDecodedErr != nil {
 		// If there is an error while decoding the hex identity public key, terminate the connection
-		return websocket.BinaryMessage, identityPublicKeyDecodedErr
+		return "", identityPublicKeyDecodedErr
+	}
+	// Get the byte sequence which was signed by the client
+	byteSequenceThatClientSigned := messageFromClient.Response.Auth.ToSign
+	// Check if the byte sequence that was signed by the client contains the initial bytes we sent to the client
+	if !bytes.HasPrefix(byteSequenceThatClientSigned, backendRandomBytes) {
```

Comment:
before this check, I would check if `ToSign` has the correct length. Just to be sure we are getting valid data.

Reply:
Agreed. Will add a check to make sure we are getting correct length.